### PR TITLE
Provide support for Open Swoole

### LINKED ISSUE #72: ext-openswoole
When mezzio-swoole will be integrated with package ext-openswoole rather than ext-swoole?

### PR DESCRIPTION
**Pull request by weierophinney**
This patch provides support for Open Swoole.

To do so, it makes the following changes:

- The package no longer declares a dependency on ext-swoole. Instead, it adds suggestions for ext-swoole and ext-openswoole, indicating one or the other should be installed.
- It modifies the `ConfigProvider` to raise an exception if NEITHER extension is loaded.
- It adds a PHPUnit bootstrap that raises an exception if NEITHER extension is loaded.

Fixes #72

### CODE REVIEW

**Comment by boesing** [comment]
Any reasons why we do not simply clone fork this package and create `mezzio-openswoole` while keeping extension requirements?
We had plenty of problems lately with missing `ext-*` requirements. That was the whole reason on why we created satellite packages for `laminas-cache`.

I personally am against this change and more open for a dedicated component rather than having a hybrid package.

**Comment by weierophinney** [comment]
> Any reasons why we do not simply clone fork this package and create `mezzio-openswoole` while keeping extension requirements?
> We had plenty of problems lately with missing `ext-*` requirements.

Yes, one really big one: currently, the two projects are 1:1 compatible, down to the classes, methods, and constants; the only difference is the actual extension name. Since Composer doesn't have a mechanism for saying "1 of any of this set of extensions is required", we're left with either (a) removing the extension requirement (what I propose here), or (b) duplicating the package with different extension requirements.

At this point, it's unclear to me if the original project is going to die and openswoole take over completely, or if they will in fact completely fork and have changes.

My issue with going the route you suggest (option (b) in my above synopsis) is that we're left with multiple locations for issues and pull requests, and we'd have to essentially port any patches from one to the other up until the point the two projects actually diverge their APIs. I think for this interim period, it'll be less complicated for consumers if they can just grab mezzio-swoole and then use whichever of the two extensions they want underneath it. Once a clear winner emerges, and/or we decide both forks are viable, we can determine whether we split the project or tighten up the requirements to reflect one extension.

**Comment by weierophinney** [comment]
@mezzio/technical-steering-committee Any other objections? (Looks like Max approved my logic for targeting both swoole and openswoole in a single repo for the time being...)

**Review by samsonasik — APPROVED** [review]

File: src/ConfigProvider.php [hunk 1]
```diff
@@ -26,7 +27,13 @@ class ConfigProvider
 {
     public function __invoke(): array
     {
-        $config = PHP_SAPI === 'cli' && extension_loaded('swoole')
+        if (! extension_loaded('swoole') && ! extension_loaded('openswoole')) {
+            throw new ExtensionNotLoadedException(
+                'One of either the swoole or openswoole extensions must be loaded to use mezzio/mezzio-swoole'
```
Comment (froschdesign) [inline src/ConfigProvider.php]: I think links to the related websites / documentation would be good here; for both extensions.

File: composer.json [hunk 2]
```diff
@@ -54,7 +54,9 @@
         "vimeo/psalm": "^4.0"
     },
     "suggest": {
-        "ext-inotify": "To use inotify based file watcher. Required for hot code reloading."
+        "ext-inotify": "To use inotify based file watcher. Required for hot code reloading.",
+        "ext-openswoole": "The package requires at least one of OpenSwoole or Swoole",
+        "ext-swoole": "The package requires at least one of OpenSwoole or Swoole"
```
Comment (froschdesign) [inline composer.json]: The trademark is "OpenSwoole" but everywhere it is written: "Open Swoole".

* https://www.swoole.co.uk
* https://github.com/openswoole
* https://github.com/mezzio/mezzio-swoole/pull/73/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R6

**Comment by froschdesign** [comment]
@weierophinney 
This component should be added to the readme file of Open Swoole: https://github.com/openswoole/swoole-src#frameworks--components

File: composer.json (hunk 2 above)
Comment (weierophinney) [reply composer.json]: I'm assuming, then, that when I refer to it in a written fashion that it should be "Open Swoole" everywhere, based on that. I've updated the patch to reflect that.

File: src/ConfigProvider.php (hunk 1 above)
Comment (weierophinney) [reply src/ConfigProvider.php]: Done.

**Comment by weierophinney** [comment]
> This component should be added to the readme file of Open Swoole: https://github.com/openswoole/swoole-src#frameworks--components

I'll do that once merged and released.

**Review by froschdesign — APPROVED** [review]

**Review by boesing — APPROVED** [review]

**Pull request merged**